### PR TITLE
Add dcos_metrics output plugin

### PR DESCRIPTION
**Pull request by branden**
https://jira.mesosphere.com/browse/DCOS_OSS-3719

DC/OS PR: https://github.com/dcos/dcos/pull/3348

### CODE REVIEW

File: Godeps [hunk 1]
```diff
@@ -71,13 +67,8 @@ github.com/rcrowley/go-metrics 1f30fe9094a513ce4c700b9a54458bbb0c96996c
 github.com/samuel/go-zookeeper 1d7be4effb13d2d908342d349d71a284a7542693
 github.com/satori/go.uuid 5bf94b69c6b68ee1b541973bb8e1144db23a194b
 github.com/shirou/gopsutil c95755e4bcd7a62bb8bd33f3a597a7c7f35e2cf3
-github.com/shirou/w32 3c9377fc6748f222729a8270fe2775d149a249ad
-github.com/Shopify/sarama 3b1b38866a79f06deddf0487d5c27ba0697ccd65
-github.com/Sirupsen/logrus 61e43dc76f7ee59a82bdf3d71033dc12bea4c77d
```
Comment (philipnrmn) [inline Godeps:76]: Additions to deps seems fine, but how come we're removing?

File: Godeps [hunk 2]
```diff
@@ -19,33 +23,26 @@ github.com/eapache/go-xerial-snappy bb955e01b9346ac19dc29eb16586c90ded99a98c
...
 github.com/go-sql-driver/mysql 2e00b5cd70399450106cec6431c2e2ce3cae5034
 github.com/gobwas/glob bea32b9cd2d6f55753d94a28e959b13f0244797a
-github.com/go-ini/ini 9144852efba7c4daf409943ee90767da62d55438
 github.com/gogo/protobuf 7b6c6391c4ff245962047fc1e2c6e08b1cdfa0e8
 github.com/golang/protobuf 8ee79997227bf9b34611aee7946ae64735e6fd93
 github.com/golang/snappy 7db9049039a047d955fe8c19b83c8ff5abd765c7
-github.com/go-ole/go-ole be49f7c07711fcb603cff39e1de7c67926dc0ba7
 github.com/google/go-cmp f94e52cad91c65a63acc1e75d4be223ea22e99bc
 github.com/gorilla/mux 53c1911da2b537f792e7cafcb446b05ffe33b996
-github.com/go-redis/redis 73b70592cdaa9e6abdfcfbf97b4a90d80728c836
```
Comment (philipnrmn) [inline Godeps:31]: the redis input plugin uses go-redis...

File: plugins/outputs/dcos_metrics/dcos_metrics.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	producer, producerChan := httpProducer.New(config)
+	d.metricChan = producerChan
+	go producer.Run()
+
+	return nil
+}
+
+func (d *DCOSMetrics) Stop() {}
```
Comment (philipnrmn) [inline plugins/outputs/dcos_metrics/dcos_metrics.go]: should we stop the producer in here, or are we OK with the process being terminated?

File: plugins/outputs/dcos_metrics/translator.go [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+	frameworkName := getAndDelete(tags, "service_name") // DC/OS services are Mesos frameworks.
+	taskName := getAndDelete(tags, "task_name")
+	executorName := getAndDelete(tags, "executor_name")
+
+	return producers.MetricsMessage{
+		Name: producers.ContainerMetricPrefix,
+		Datapoints: datapointsFromMetric(m, map[string]string{
+			"container_id":  containerID,
+			"executor_name": executorName,
+			// TODO(branden): These are set by createContainerDatapoints() in dcos-metrics, but I'm not sure where to get these values. Should these be provided by a processor plugin?
```
Comment (philipnrmn) [inline plugins/outputs/dcos_metrics/translator.go]: these can be omitted.

File: plugins/outputs/dcos_metrics/dcos_metrics.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		listenHost, listenPort, err = splitHostPort(d.Listen)
+		if err != nil {
+			return httpProducer.Config{}, errors.New(fmt.Sprintf("error reading listen: %s", err))
+		}
+	}
+
+	switch d.DCOSNodeRole {
+	case "master", "agent":
+	default:
+		return httpProducer.Config{}, errors.New("error reading dcos_node_role: must be one of master or agent")
```
Comment (philipnrmn) [inline plugins/outputs/dcos_metrics/dcos_metrics.go:112]: could this be agent_public?

**Review by philipnrmn — COMMENTED** [review]
Just a few questions

File: Godeps (hunk 1 above)
Comment (branden) [reply Godeps:76]: `gdm save` made all the edits to this file. Most of these lines are just reordered, although there are a few removed. CI is able to build Telegraf after restoring deps from this file, so I assume the removed deps weren't being used.

File: Godeps (hunk 2 above)
Comment (branden) [reply Godeps:31]: `gdm save` relocated this line.

File: plugins/outputs/dcos_metrics/dcos_metrics.go (hunk 3 above)
Comment (branden) [reply plugins/outputs/dcos_metrics/dcos_metrics.go]: The producer doesn't offer a mechanism to stop it. I'm not sure if there's anything to clean up anyway, `dcos-metrics` doesn't seem to do anything special to shut down. I'll add a comment here to make clear that we're deliberately doing nothing.

File: plugins/outputs/dcos_metrics/dcos_metrics.go (hunk 5 above)
Comment (branden) [reply plugins/outputs/dcos_metrics/dcos_metrics.go:112]: `dcos-metrics` only accepts master and agent: https://github.com/dcos/dcos-metrics/blob/aa08dd7bd88e76dcc67d0171ea0ae30c5b818d83/config.go#L246-L252

**Review by philipnrmn — DISMISSED** [review]

**Review by philipnrmn — APPROVED** [review]

**Pull request merged**